internal/util/apiclient/websocket_log_reader.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2024 Daytona Platforms Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package apiclient
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/daytonaio/daytona/cmd/daytona/config"
+	"github.com/daytonaio/daytona/pkg/logs"
+	logs_view "github.com/daytonaio/daytona/pkg/views/logs"
+	"github.com/gorilla/websocket"
+)
+
+var workspaceLogsStarted bool
+
+func ReadWorkspaceLogs(activeProfile config.Profile, workspaceId string, projectNames []string, stopLogs *bool) {
+	var wg sync.WaitGroup
+	query := "follow=true"
+
+	logs_view.CalculateLongestPrefixLength(projectNames)
+
+	for index, projectName := range projectNames {
+		wg.Add(1)
+		go func(projectName string) {
+			defer wg.Done()
+
+			for {
+				// Make sure workspace logs started before showing any project logs
+				if !workspaceLogsStarted {
+					time.Sleep(250 * time.Millisecond)
+					continue
+				}
+
+				ws, _, err := GetWebsocketConn(fmt.Sprintf("/log/workspace/%s/%s", workspaceId, projectName), &activeProfile, &query)
+				// We want to retry getting the logs if it fails
+				if err != nil {
+					// TODO: return log.Trace once https://github.com/daytonaio/daytona/issues/696 is resolved
+					// log.Trace(apiclient_util.HandleErrorResponse(res, err))
+					time.Sleep(500 * time.Millisecond)
+					continue
+				}
+
+				readJSONLog(ws, stopLogs, index)
+				ws.Close()
+				break
+			}
+		}(projectName)
+	}
+
+	for {
+		ws, _, err := GetWebsocketConn(fmt.Sprintf("/log/workspace/%s", workspaceId), &activeProfile, &query)
+		// We want to retry getting the logs if it fails
+		if err != nil {
+			// TODO: return log.Trace once https://github.com/daytonaio/daytona/issues/696 is resolved
+			// log.Trace(apiclient_util.HandleErrorResponse(res, err))
+			time.Sleep(250 * time.Millisecond)
+			continue
```

Comment:
Is there an infinite loop possibility?

Reply:
This is the log reading code we used before as well. It runs `GetWebsocketConn` each loop and waits 250ms on fail.

This is called from the client so it can only last as long as the underlying process (workspace creation) which the user can stop at any time so I wouldn't say there's an infinite loop possibility. Even if it really errors every iteration, that would still be the "inteded" sequence because we want to keep trying to ping the logs websocket for however long creation lasts.